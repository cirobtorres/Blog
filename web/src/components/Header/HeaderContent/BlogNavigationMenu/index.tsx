"use client";

import * as React from "react";
import Link from "next/link";
import {
  NavigationMenu,
  NavigationMenuItem,
  NavigationMenuLink,
  NavigationMenuList,
  navigationMenuTriggerStyle,
} from "../../../Shadcnui/navigation-menu";
import { usePathname } from "next/navigation";

const BlogNavigationMenu = () => {
  const pathname = usePathname();

  const getCleanPath = (path: string) => {
    const url = new URL(path, "http://example.com");
    return url.pathname;
  };

  const path = getCleanPath(pathname);
  const isArticlesActive = path === "/artigos" || path.startsWith("/artigos/");
  const isAboutActive = path === "/sobre" || path.startsWith("/sobre/");

  return (
    <NavigationMenu
      data-testid="blog-navigation-menu"
      className="max-[1000px]:hidden"
    >
      <NavigationMenuList>
        <NavigationMenuItem>
          <Link href="/artigos" legacyBehavior passHref>
            <NavigationMenuLink
              className={`text-sm w-20 ${
                isArticlesActive ? "text-blog-foreground-highlight " : ""
              } ${navigationMenuTriggerStyle()}`}
            >
              Artigos
            </NavigationMenuLink>
          </Link>
        </NavigationMenuItem>
        <NavigationMenuItem>
          <Link href="/sobre" legacyBehavior passHref>
            <NavigationMenuLink
              className={`text-sm w-20 ${
                isAboutActive ? "text-blog-foreground-highlight " : ""
              } ${navigationMenuTriggerStyle()}`}
            >
              Sobre
            </NavigationMenuLink>
          </Link>
        </NavigationMenuItem>
      </NavigationMenuList>
    </NavigationMenu>
  );
};

export default BlogNavigationMenu;
