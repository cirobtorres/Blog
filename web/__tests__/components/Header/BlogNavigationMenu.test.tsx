import { usePathname } from "next/navigation";
import { cleanup, render, screen } from "@testing-library/react";
import BlogNavigationMenu from "../../../src/components/Header/HeaderContent/BlogNavigationMenu";

jest.mock("next/navigation", () => ({
  // HeaderContent contains multiple nested components.
  // BlogNavigationMenu uses the usePathname hook, so we need to mock it.
  // SearchBar uses the useSearchParams hook, so we need to mock it as well.
  usePathname: jest.fn(() => "/"),
  useSearchParams: jest.fn(() => new URLSearchParams()),
}));

describe("BlogNavigationMenu", () => {
  it("renders blog navigation menu", () => {
    render(<BlogNavigationMenu />);
    const blogNavMenu = screen.getByTestId("blog-navigation-menu");
    expect(blogNavMenu).toBeInTheDocument();
  });

  it('should highlight "Artigos" for nested article paths', () => {
    const testPaths = [
      "/artigos",
      "/artigos/",
      "/artigos/react-testing",
      "/artigos/12345",
      "/artigos?page=2",
    ];

    testPaths.forEach((path) => {
      (usePathname as jest.Mock).mockReturnValue(path);
      render(<BlogNavigationMenu />);
      const articlesItem = screen.getByRole("link", { name: /Artigos/i });
      expect(articlesItem).toHaveClass("text-blog-foreground-highlight");
      cleanup();
    });
  });

  it('should not highlight "Artigos" for unrelated paths', () => {
    const testPaths = ["/artigo", "/artigosss", "/sobre", "/", "/contato"];

    testPaths.forEach((path) => {
      (usePathname as jest.Mock).mockReturnValue(path);
      render(<BlogNavigationMenu />);
      const articlesItem = screen.getByRole("link", { name: /Artigos/i });
      expect(articlesItem).not.toHaveClass("text-blog-foreground-highlight");
      cleanup();
    });
  });

  it('should highlight "Sobre" for nested about paths', () => {
    const testPaths = ["/sobre", "/sobre/", "/sobre/equipe", "/sobre/historia"];

    testPaths.forEach((path) => {
      (usePathname as jest.Mock).mockReturnValue(path);
      render(<BlogNavigationMenu />);
      const aboutItem = screen.getByRole("link", { name: /Sobre/i });
      expect(aboutItem).toHaveClass("text-blog-foreground-highlight");
      cleanup();
    });
  });

  it('should not highlight "Sobre" for unrelated paths', () => {
    const testPaths = ["/sobr", "/sobreee", "/artigos", "/", "/contato"];

    testPaths.forEach((path) => {
      (usePathname as jest.Mock).mockReturnValue(path);
      render(<BlogNavigationMenu />);
      const aboutItem = screen.getByRole("link", { name: /Sobre/i });
      expect(aboutItem).not.toHaveClass("text-blog-foreground-highlight");
      cleanup();
    });
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(<BlogNavigationMenu />);
    expect(asFragment()).toMatchSnapshot();
  });
});
