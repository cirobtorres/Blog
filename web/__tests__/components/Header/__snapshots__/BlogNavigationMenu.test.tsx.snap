// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogNavigationMenu matches the snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="Main"
    class="relative z-10 flex max-w-max flex-1 items-center justify-center max-[1000px]:hidden"
    data-orientation="horizontal"
    data-testid="blog-navigation-menu"
    dir="ltr"
  >
    <div
      style="position: relative;"
    >
      <ul
        class="group flex flex-1 list-none items-center justify-center space-x-1"
        data-orientation="horizontal"
        dir="ltr"
      >
        <li>
          <a
            class="text-sm w-20  py-2 inline-flex items-center justify-center rounded transition-colors duration-500 hover:text-blog-foreground-readable-hover focus:text-blog-foreground-readable-hover focus-visible:outline focus-visible:outline-2 focus-visible:outline-blog-foreground-readable-hover disabled:pointer-events-none focus-visible:bg-[hsl(0,0%,14.9%,0.75)] disabled:opacity-50 data-[active]:text-blog-foreground-readable-hover hover:bg-[hsl(0,0%,14.9%,0.75)] data-[state=open]:text-blog-foreground-readable-hover data-[state=open]:bg-[hsl(0,0%,14.9%,0.75)] group"
            data-radix-collection-item=""
            href="/artigos"
          >
            Artigos
          </a>
        </li>
        <li>
          <a
            class="text-sm w-20  py-2 inline-flex items-center justify-center rounded transition-colors duration-500 hover:text-blog-foreground-readable-hover focus:text-blog-foreground-readable-hover focus-visible:outline focus-visible:outline-2 focus-visible:outline-blog-foreground-readable-hover disabled:pointer-events-none focus-visible:bg-[hsl(0,0%,14.9%,0.75)] disabled:opacity-50 data-[active]:text-blog-foreground-readable-hover hover:bg-[hsl(0,0%,14.9%,0.75)] data-[state=open]:text-blog-foreground-readable-hover data-[state=open]:bg-[hsl(0,0%,14.9%,0.75)] group"
            data-radix-collection-item=""
            href="/sobre"
          >
            Sobre
          </a>
        </li>
      </ul>
    </div>
    <div
      class="absolute left-0 top-full flex justify-center"
    />
  </nav>
</DocumentFragment>
`;
